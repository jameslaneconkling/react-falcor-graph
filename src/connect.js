/* eslint-disable max-len, no-underscore-dangle */
const {
  hoistStatics,
  compose,
  setDisplayName,
  wrapDisplayName,
  mapPropsStream
} = require('recompose');
const {
  Observable
} = require('rxjs/Observable');
require('rxjs/add/observable/from');
require('rxjs/add/operator/distinctUntilChanged');
require('rxjs/add/operator/merge');
require('rxjs/add/operator/withLatestFrom');
require('rxjs/add/operator/let');
require('rxjs/add/operator/switchMap');
require('rxjs/add/operator/startWith');
require('rxjs/add/operator/map');
require('rxjs/add/operator/last');
require('rxjs/add/operator/catch');
require('rxjs/add/operator/auditTime');
require('rxjs/add/operator/delay');
var {
  animationFrame
} = require('rxjs/scheduler/animationFrame');


/**
 * @param {(props) => paths | paths} paths
 * @param {Object} falcorModel
 * @param {Observable<>} change$
 * @param {Object} options
 */
const connectFalcorStream = exports.connectFalcorStream = (
  paths,
  falcorModel,
  change$,
  {
    errorHandler = (props, error) => Observable.of(Object.assign({}, props, { graphFragment: {}, graphFragmentStatus: 'error', error })),
    prefixStream = props$ => props$
  } = {}
) => {
  const _models = {};

  return (props$) => {
    const _props$ = Observable.from(props$);

    const graphQueryResponse$ = _props$
      // .distinctUntilChanged((prev, next) => equals(
      //   typeof paths === 'function' ? paths(prev) : paths,
      //   typeof paths === 'function' ? paths(next) : paths
      // ))
      .merge(Observable.from(change$).withLatestFrom(_props$, (_, props) => props))
      .let(prefixStream)
      .switchMap((props) => {
        const _paths = typeof paths === 'function' ? paths(props) : paths;

        if (_paths === null) {
          return Observable.of(Object.assign({}, props, { graphFragment: {}, graphFragmentStatus: 'complete' }));
        }

        let model;
        if (!falcorModel._recycleJSON) {
          model = falcorModel;
        } else if (!_models[props.id]) {
          model = _models[props.id] = falcorModel._clone({ _seed: {} });
        } else {
          model = _models[props.id];
        }

        const graphQuery$ = Observable.from(model.get(..._paths).progressively());

        return graphQuery$
          // .distinctUntilKeyChanged('version') or .distinctUntilChanged((prev, next) => prev.graphFrag)
          .map(graphFragment => ({ graphFragment, graphFragmentStatus: 'next' }))
          .merge(graphQuery$.last().map(graphFragment => ({ graphFragment, graphFragmentStatus: 'complete' })))
          .catch((err, caught) => errorHandler(props, err, caught));
      });

    return _props$
      .map(props => Object.assign({}, props, { graphFragment: {}, graphFragmentStatus: 'next' }))
      .merge(graphQueryResponse$.withLatestFrom(_props$, (graphQuery, props) => Object.assign({}, props, graphQuery)))
      .auditTime(0, animationFrame);
  };
};


/**
 * @param {(props) => paths | paths} paths
 * @param {Object} falcorModel
 * @param {Observable<>} change$
 * @param {Object} options
 */
exports.default = (paths, falcorModel, change$, options) => WrappedComponent =>
  hoistStatics(compose(
    setDisplayName(wrapDisplayName(WrappedComponent, 'FalcorConnect')),
    mapPropsStream(connectFalcorStream(paths, falcorModel, change$, options))
  ))(WrappedComponent);
