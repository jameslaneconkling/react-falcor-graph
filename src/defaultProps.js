require('rxjs/add/operator/map');


exports.default = projection => props$ =>
  props$.map((props) => Object.assign({}, projection(props), props));
