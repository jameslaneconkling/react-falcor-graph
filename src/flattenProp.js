require('rxjs/add/operator/map');


exports.default = propName => props$ =>
  props$.map(props => Object.assign({}, props, props[propName]));
