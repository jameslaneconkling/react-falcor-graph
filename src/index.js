exports.default = require('./connect');
exports.withGraphFragment = require('./connect').withGraphFragment;
