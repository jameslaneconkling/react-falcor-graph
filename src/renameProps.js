require('rxjs/add/operator/map');
const { mapKeys } = require('./utils');


exports.default = nameMap => props$ =>
  props$.map(props => mapKeys((key) => nameMap[key] !== undefined ? nameMap[key] : key , props));
