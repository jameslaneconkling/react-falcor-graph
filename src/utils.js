exports.mapHashMap = (predicate, hashMap) => {
  let out = {};
  let hashMapKeys = Object.keys(hashMap);

  for (let i = 0; i < hashMapKeys.length; i++) {
    out[hashMapKeys[i]] = predicate(hashMap[hashMapKeys[i]]);
  }

  return out;
};


const range = (from = 0, to) => {
  const list = [];
  while (from <= to) {
    list.push(from);
    from += 1;
  }
  return list;
};


const expandPath = exports.expandPath = pathOrPathSet =>
  pathOrPathSet.reduce((paths, keyOrKeySet) => {
    if (Array.isArray(keyOrKeySet)) {
    // keySet, e.g. ['name', 'age']
      return keyOrKeySet.reduce((expandedPaths, key) =>
        [...expandedPaths, ...paths.map(path => [...path, key])],
      []);
    } else if (typeof keyOrKeySet === 'object') {
    // range, e.g. [{ from: 1, to: 3 }]
      return range(keyOrKeySet.from, keyOrKeySet.to).reduce((expandedPaths, key) =>
        [...expandedPaths, ...paths.map(path => [...path, key])],
      []);
    }
    // key
    return paths.map(path => [...path, keyOrKeySet]);
  }, [[]]);


exports.expandPaths = paths =>
  paths.reduce((expandedPaths, path) => [...expandedPaths, ...expandPath(path)], []);


const walkTree = exports.walkTree = (path, tree, graph = tree) => {
  if (path.length === 0) {
    return tree;
  } else if (path.length === 1) {
    return tree[path[0]];
  } else if (path[0] in tree && tree[path[0]].$type === 'ref') {
    // if encountering a ref, go back to the root graph and evaluate new path from there
    return walkTree([...tree[path[0]].value, ...path.slice(1)], graph);
  } else if (path[0] in tree) {
    return walkTree(path.slice(1), tree[path[0]], graph);
  }
  return undefined;
};


exports.isPaths = paths => {
  if (!Array.isArray(paths)) {
    return false;
  }

  if (paths.length === 0) {
    return false;
  }

  for (let i = 0; i < paths.length; i++) {
    if (!Array.isArray(paths[i])) {
      return false;
    }
  }

  return true;
};


exports.compose = (...fns) => {
  if (fns.length === 0) {
    return args => args;
  } else if (fns.length === 1) {
    return fns[0];
  }

  return fns.reduce((a, b) => (...args) => a(b(...args)));
};
