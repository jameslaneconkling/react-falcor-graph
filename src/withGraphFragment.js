const {
  Observable
} = require('rxjs/Observable');
require('rxjs/add/observable/from');
require('rxjs/add/observable/empty');
require('rxjs/add/operator/withLatestFrom');
require('rxjs/add/operator/last');
require('rxjs/add/operator/merge');
require('rxjs/add/operator/repeatWhen');
require('rxjs/add/operator/let');
require('rxjs/add/operator/switchMap');
require('rxjs/add/operator/startWith');
require('rxjs/add/operator/map');
require('rxjs/add/operator/catch');
require('rxjs/add/operator/auditTime');


/**
 * @param {(props) => paths | paths | null | Error} paths
 * @param {Object} falcorModel
 * @param {Observable<>} graphChange$
 * @param {Object} options
 */
exports.default = (
  paths,
  falcorModel,
  graphChange$,
  {
    errorHandler = error => Observable.of({ graphFragment: {}, graphFragmentStatus: 'error', error }),
    prefixStream = props$ => props$,
    auditTime = 0
  } = {}
) => {
  const _models = {};

  return (props$) => {
    const _props$ = Observable.from(props$);

    const graphQueryResponse$ = _props$
      .let(prefixStream)
      .switchMap((props) => {
        const _paths = typeof paths === 'function' ? paths(props) : paths;

        if (!_paths) {
          return Observable.of(Object.assign({}, props, { graphFragment: {}, graphFragmentStatus: 'complete' }));
        } else if (paths instanceof Error) {
          return Observable.of(Object.assign({}, props, { graphFragment: {}, graphFragmentStatus: 'error', error: paths.message }));
        }

        let model;
        if (!falcorModel._recycleJSON) {
          model = falcorModel;
        } else if (!_models[props.id]) {
          model = _models[props.id] = falcorModel._clone({ _seed: {} });
        } else {
          model = _models[props.id];
        }

        const graphQuery$ = Observable.from(model.get(..._paths).progressively());

        return graphQuery$
          .map(graphFragment => ({ graphFragment, graphFragmentStatus: 'next' }))
          .catch((err, caught) => errorHandler(err, props, caught))
          .merge(graphQuery$
            .last()
            .map(graphFragment => ({ graphFragment, graphFragmentStatus: 'complete' }))
            .catch(() => Observable.empty())
          )
          .repeatWhen(() => graphChange$);
      });

    return props$
      .map(props => Object.assign({}, props, { graphFragment: {}, graphFragmentStatus: 'next' }))
      .merge(graphQueryResponse$.withLatestFrom(props$, (response, props) => Object.assign({}, props, response)))
      .auditTime(auditTime);

    // Note - the more straightforward use of combineLatest fails when props$ fires rapidly
    // a situation with no test case yet
    // return Observable.combineLatest(
    //   _props$,
    //   graphQueryResponse$.startWith({ graphFragment: {}, graphFragmentStatus: 'next' }),
    //   (props, graphQueryResponse) => Object.assign({}, props, graphQueryResponse)
    // );
  };
};
