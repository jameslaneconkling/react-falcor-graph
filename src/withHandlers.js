const { mapHashMap } = require('./utils');
require('rxjs/add/operator/map');


exports.default = (handlers) => {
  let _props;

  const cachedHandlers = mapHashMap(handlers, handler => (...args) => handler(_props)(...args));

  return props$ => props$
    .map((props) => {
      _props = props;
      return Object.assign({}, props, cachedHandlers);
    });
};
