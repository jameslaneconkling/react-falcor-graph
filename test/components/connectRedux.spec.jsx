/* eslint-disable react/prop-types */
import React                from 'react';
import renderer             from 'react-test-renderer';
import {
  createStore
}                           from 'redux';
import { connectRedux }     from '../../src';


test('connectRedux render component with ', () => {
  const App = ({ id, item }) => (
    <div>Id: {id} - {item.title}</div>
  );

  const store = createStore((state = {
    items: {
      a: { title: 'Item A' }
    }
  }) => state);
  const mapState = (state, { id }) => ({ item: state.items[id] });
  const mapDispatch = () => ({});

  const ConnectedApp = connectRedux(
    store, mapState, mapDispatch
  )(App);


  const tree = renderer.create(<ConnectedApp id="a" />).toJSON();

  expect(tree).toMatchSnapshot();
});
