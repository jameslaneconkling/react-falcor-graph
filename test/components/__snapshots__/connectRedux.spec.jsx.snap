// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectRedux renders component with store data 1`] = `
<div>
  Id: 
  a
   - 
  Item A
</div>
`;
